### DISCUSSION TITLE
Update PCB instead of recreate?

### DISCUSSION THREAD

**Discussion by shanemmattner** [xesscorp/skidl#161] [Ideas]
Is there a way to update a PCB with new parts instead of regenerating the entire thing?  I ran into this situation where I create a circuit and route it.  Then when I add a part and regenerate the PCB the parts are regenerated and all my work is gone.

**Comment by devbisme** [comment]
There's currently no way to update the existing PCB file. Your best bet is to create the netlist using SKiDL `generate_netlist()` and then import it into your PCB file.

But here's the problem: if you insert a new part into your SKiDL code, then that's going to cause a renumbering of parts. Then PCBNEW is going to have a problem matching parts from the netlist with existing parts in the PCB. KiCad gets around this problem when schematic symbols are renumbered by maintaining an immutable key for each schematic symbol that never changes. Then you can do a netlist import using this key instead of the symbol numbers. This technique isn't possible with SKiDL because there is no way to persistently store part numbers or keys: it gets regenerated each time the SKiDL code is executed.

A potential solution to this is to build a tool that:
1. Extracts the netlist from the PCB file.
2. Compares it to the netlist generated by SKiDL.
3. Generates a mapping from the part numbers in the PCB netlist to the part numbers in the SKiDL netlist (probably using some graph isomorphism techniques but I'm handwaving here).
4. Renumbers the parts in the SKiDL netlist.

So it's a pending project.
